include/kb_client.hpp

```diff
@@ -59,194 +59,228 @@ std::map<msgpack::type::object_type, std::string> msgpack_type_map = {
+     * CastErrorMsgpackObject: if cast fails
      */
     template<typename T>
     T as() const {
-        checkTypeByIndex<T>(value_.type);
         try {
             return value_.as<T>();
         } catch(std::bad_cast& e) {
-            throw CastErrorMsgpack();
+            std::string error_msg;
+            if (size_)
+                error_msg = ("The expected type is a(n) "
+                             + containerType() + " of " + dtype());
+            else
+                error_msg = ("The expected type is " + dtype());
+
+            throw CastErrorMsgpackObject(error_msg);
         }
     }
 
-    std::string dtype() const { return dtype_; }
+    std::string dtype() const override { return dtype_; }
 
-    std::size_t size() const { return size_; }
+    std::size_t size() const override { return size_; }
 
-    std::string subType() const {
-        if (!size_) return "";  // NIL
-        if (size_ == 1) return dtype_;
-        if (dtype_ == "MSGPACK_OBJECT_ARRAY") return msgpack_type_map.at(value_.via.array.ptr[0].type);
-        // TODO: should I give "char" or "unsigned char"?
-        if (dtype_ == "MSGPACK_OBJECT_BIN") return "byte";
-        return "unimplemented";
+    std::vector<std::size_t> shape() const override {
+        if (size_) return std::vector<std::size_t>({size_});
+        return std::vector<std::size_t>();
+    }
+
+    std::string containerType() const override {
+        if (!size_) return ""; // scalar and NIL
+        if (value_.type == msgpack::type::object_type::ARRAY ||
+                value_.type == msgpack::type::object_type::BIN)
+            return "array-like";
+        if (value_.type == msgpack::type::object_type::MAP) return "map";
+        return msgpack_type_map.at(value_.type);
     }
 };
 
+using ObjectMap = std::map<std::string, MsgpackObject>;
+using ObjectPair = std::pair<std::string, MsgpackObject>;
+
+namespace detail {
+
+template<typename Container, typename ElementType>
+struct as_imp {
+    Container operator()(void* ptr_, std::size_t size) {
+        auto ptr = reinterpret_cast<const ElementType*>(ptr_);
+        return Container(ptr, ptr + size);
+    }
+};
+
+// partial specialization for std::array
+template<typename ElementType, std::size_t N>
+struct as_imp<std::array<ElementType, N>, ElementType> {
+    std::array<ElementType, N> operator()(void*ptr_, std::size_t size) {
+        if (size != N)
+            throw CastErrorArray("The input size " + std::to_string(N)
+                                 + " is different from the expected size "
+                                 + std::to_string(size));
+        auto ptr = reinterpret_cast<const ElementType*>(ptr_);
+        std::array<ElementType, N> arr;
+        memcpy(arr.data(), ptr, size * sizeof(ElementType));
+        return arr;
+    }
+};
+
+}  // detail
+
 /*
  * A container held a pointer to the data chunk and other useful information.
  */
-class Array {
+class Array : public Object {
```

Comment:
Maybe we can call this `NDArray` to fit with the karabo framework, what do you think?